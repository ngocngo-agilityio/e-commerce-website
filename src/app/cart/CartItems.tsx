'use client';

// Libs
import { useCallback, useMemo } from 'react';
import { Box, Flex, Heading, Text, Divider, Button } from '@chakra-ui/react';

// Apis
import { updateCartItemQuantity, removeCartItem } from '@apis';

// Hooks
import { useCustomToast } from '@hooks';

// Stores
import { useCartStore } from '@stores';

// Utils
import { formatCurrency, calculateTotalPrice } from '@utils';

// Components
import { CartTable } from '@components';

const CartItems = (): JSX.Element => {
  const { showToast } = useCustomToast();

  // Cart store
  const cartItems = useCartStore((state) => state.cartItems);

  const total = useMemo(() => calculateTotalPrice(cartItems), [cartItems]);
  const formattedTotal = useMemo(() => formatCurrency(total), [total]);

  const handleRemoveCartItem = useCallback(
    async (cartId: string) => {
      const { error } = await removeCartItem(cartId);

      if (error) {
        showToast(error);
      }
    },
    [showToast],
  );

  // TODO: Update later
  const handleCheckout = () => {};

  const handleChangeQuantity = useCallback(
    async (cartId: string, quantity: number) => {
      const { error } = await updateCartItemQuantity({ quantity, cartId });

      if (error) {
        showToast(error);
      }
    },
    [showToast],
  );

  return (
    <>
      <CartTable
        cart={cartItems}
        onCartItemRemove={handleRemoveCartItem}
        onQuantityChange={handleChangeQuantity}
      />
      <Heading mt="83px" mb="38px">
        Cart Totals
      </Heading>
      <Box w="531px" mb="41px">
        <Flex justifyContent="space-between" py="17px">
          <Text>Subtotal</Text>
          <Text>{formattedTotal}</Text>
        </Flex>
        <Divider />
        <Flex justifyContent="space-between" py="17px">
          <Text>ShippingFree</Text>
          <Text>FREE!!!</Text>
        </Flex>
        <Divider />
        <Flex justifyContent="space-between" py="17px">
          <Text fontFamily="arimo" fontWeight="bold">
            Total
          </Text>
          <Text>{formattedTotal}</Text>
        </Flex>
      </Box>

      <Button variant="checkoutBtn" fontSize="20px" w="262px" h="58px" onClick={handleCheckout}>
        Proceed to Checkout
      </Button>
    </>
  );
};

export default CartItems;
