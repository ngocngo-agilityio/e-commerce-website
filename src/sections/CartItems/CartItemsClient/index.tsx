'use client';

// Libs
import { useMemo } from 'react';
import { Box, Flex, Heading, Text, Divider, Button } from '@chakra-ui/react';

// Types
import { CartItem } from '@types';

// Utils
import { calculateTotalPrice, formatCurrency } from '@utils';

// Components
import { CartTable } from '@components';

interface CartItemClient {
  cartItems: CartItem[];
}

const CartItemsClient = ({ cartItems }: CartItemClient): JSX.Element => {
  const total = useMemo(() => calculateTotalPrice(cartItems), [cartItems]);
  const formattedTotal = formatCurrency(total);

  // TODO: Update later
  const handleCheckout = () => {};

  // TODO: Update later
  const handleRemoveCartItem = () => {};

  // TODO: Update later
  const handleChangeQuantity = () => {};

  return (
    <Box mt="70px">
      <CartTable
        cart={cartItems}
        onCartItemRemove={handleRemoveCartItem}
        onQuantityChange={handleChangeQuantity}
      />
      <Heading mt="83px" mb="38px">
        Cart Totals
      </Heading>
      <Box w="531px" mb="41px">
        <Flex justifyContent="space-between" py="17px">
          <Text>Subtotal</Text>
          <Text>{formattedTotal}</Text>
        </Flex>
        <Divider />
        <Flex justifyContent="space-between" py="17px">
          <Text>ShippingFree</Text>
          <Text>FREE!!!</Text>
        </Flex>
        <Divider />
        <Flex justifyContent="space-between" py="17px">
          <Text fontFamily="arimo" fontWeight="bold">
            Total
          </Text>
          <Text>{formattedTotal}</Text>
        </Flex>
      </Box>

      <Button variant="checkoutBtn" w="262px" h="58px" onClick={handleCheckout}>
        Proceed to Checkout
      </Button>
    </Box>
  );
};

export default CartItemsClient;
