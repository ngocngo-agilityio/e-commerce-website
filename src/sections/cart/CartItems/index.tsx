// Apis
import { getCartItems } from '@apis';

// Sections
import CartItemsClient from '../CartItemsClient';

const CarItems = async (): Promise<JSX.Element> => {
  const { data } = await getCartItems();

  return <CartItemsClient cartItems={data} />;
};

export default CarItems;
