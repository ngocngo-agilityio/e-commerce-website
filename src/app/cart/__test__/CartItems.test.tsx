// Libs
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

// Apis
import { removeCartItem, updateCartItemQuantity } from '@apis';

// Stores
import { useCartStore } from '@stores';

// Mocks
import { MOCK_CART_ITEMS } from '@mocks';

// Components
import CartItems from '../CartItems';

const mockShowToast = jest.fn();
jest.mock('@apis');
jest.mock('@hooks', () => {
  const originalModule = jest.requireActual('@hooks');

  return {
    __esModule: true,
    ...originalModule,
    useCustomToast: jest.fn(() => ({ showToast: mockShowToast })),
  };
});

describe('CartItems', () => {
  beforeEach(() => {
    useCartStore.setState({
      cartItems: [...MOCK_CART_ITEMS],
    });
  });

  test('should render CartItems successfully', () => {
    const { container } = render(<CartItems />);

    expect(container).toBeInTheDocument();
  });

  test('should match snapshot for CartItems', () => {
    const { container } = render(<CartItems />);

    expect(container).toMatchSnapshot();
  });

  test('handle remove cart item', async () => {
    const expectedResponse = {
      error: 'Remove cart item failed',
    };

    (
      removeCartItem as jest.MockedFunction<typeof removeCartItem>
    ).mockResolvedValue(expectedResponse);

    render(<CartItems />);

    fireEvent.click(screen.getByText('x'));

    await waitFor(() =>
      fireEvent.click(screen.getByRole('button', { name: 'Yes' })),
    );

    expect(mockShowToast).toHaveBeenCalled();
  });

  test('handle change quantity', async () => {
    const expectedResponse = {
      error: 'Change quantity cart item failed.',
    };

    (
      updateCartItemQuantity as jest.MockedFunction<
        typeof updateCartItemQuantity
      >
    ).mockResolvedValue(expectedResponse);

    render(<CartItems />);

    await waitFor(() =>
      fireEvent.click(screen.getByRole('button', { name: '+' })),
    );

    expect(mockShowToast).toHaveBeenCalled();
  });
});
