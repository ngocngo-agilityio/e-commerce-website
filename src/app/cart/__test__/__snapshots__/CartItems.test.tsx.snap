// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItems should match snapshot for CartItems 1`] = `
<div>
  <table
    class="chakra-table css-0"
  >
    <thead
      class="css-0"
    >
      <tr
        class="css-0"
      >
        <th
          class="css-0"
        />
        <th
          class="css-0"
        />
        <th
          class="css-0"
        >
          Product
        </th>
        <th
          class="css-0"
        >
          Price
        </th>
        <th
          class="css-0"
        >
          Quantity
        </th>
        <th
          class="css-0"
        >
          Total
        </th>
      </tr>
    </thead>
    <tbody
      class="css-0"
    >
      <tr
        class="css-0"
      >
        <td
          class="css-1a88wtd"
        >
          <div
            class="css-ie49k9"
          >
            x
          </div>
        </td>
        <td
          class="css-0"
        >
          <div
            class="css-1bdr4yj"
          >
            <img
              alt="Product Image"
              data-nimg="fill"
              decoding="async"
              fetchpriority="high"
              sizes="100vw"
              src="/_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=3840&q=75"
              srcset="/_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Ffirebasestorage.googleapis.com%2Fv0%2Fb%2Fecommerce-fashion-16e2e.appspot.com%2Fo%2Fblack-polo-shirt.webp%3Falt%3Dmedia&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
            />
          </div>
        </td>
        <td
          class="css-0"
          style="font-weight: bold;"
        >
          <a
            href="/products/1"
          >
            Plain White Shirt
          </a>
        </td>
        <td
          class="css-0"
        >
          $69.00
        </td>
        <td
          class="css-0"
        >
          <div
            class="css-wpwawf"
          >
            <div
              class="css-1en42kf"
            >
              <button
                class="chakra-button css-15ey06b"
                type="button"
              >
                -
              </button>
            </div>
            <div
              class="css-1en42kf"
            >
              <span
                class="chakra-text css-uh9d7y"
              >
                2
              </span>
            </div>
            <div
              class="css-1en42kf"
            >
              <button
                class="chakra-button css-d83nk5"
                type="button"
              >
                +
              </button>
            </div>
          </div>
        </td>
        <td
          class="css-0"
        >
          $138.00
        </td>
      </tr>
    </tbody>
  </table>
  <h2
    class="chakra-heading css-1k51fm5"
  >
    Cart Totals
  </h2>
  <div
    class="css-3u1iq8"
  >
    <div
      class="css-19y425s"
    >
      <p
        class="chakra-text css-0"
      >
        Subtotal
      </p>
      <p
        class="chakra-text css-0"
      >
        $138.00
      </p>
    </div>
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-1upb9tn"
    />
    <div
      class="css-19y425s"
    >
      <p
        class="chakra-text css-0"
      >
        ShippingFree
      </p>
      <p
        class="chakra-text css-0"
      >
        FREE!!!
      </p>
    </div>
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-1upb9tn"
    />
    <div
      class="css-19y425s"
    >
      <p
        class="chakra-text css-19q2mny"
      >
        Total
      </p>
      <p
        class="chakra-text css-0"
      >
        $138.00
      </p>
    </div>
  </div>
  <button
    class="chakra-button css-70vipg"
    type="button"
  >
    Proceed to Checkout
  </button>
</div>
`;
